README.md

Matlab MEX interface to the wholeBodyModel C++ interface, implemented to be able to create forward dynamics on matlab for prototyping purposes.

## Installation
It is recommended to install the `mex-wholebodymodel` throught the [`codyco-superbuild`](https://github.com/robotology/codyco-superbuild/).
Once you have installed the `codyco-superbuild`, the `mex-wholebodymodel` files should have been installed in
`${CODYCO_SUPERBUILD_ROOT}/build/install/mex` and its subdirectories (for the `mex-wholebodymodel`, this directories are `mexwbi-wrappers` and `mexwbi-utilities`).
To execute scripts that use `mex-wholebodymodel`, make sure that this directories are part of your [MATLAB search path](http://www.mathworks.com/help/matlab/ref/path.html). 
To this purpose, run only once the script `startup_mexWholeBodyModel.m` in `${CODYCO_SUPERBUILD_ROOT}/build/install/share/mexWholeBodyModel`. This should be enough to premanently 
add the required directories to your MATLAB path.

## Examples 

### Rigid Body Dynamics
An example on how to use mex-wholebodymodel to compute the dynamics quantities of 
a rigid body is available at [examples/rigidBodyDynamics.m](examples/rigidBodyDynamics.m).

## Controllers
In [controllers](/controllers) folder the user can find different whole-body controllers implemented using the mex-wholebodymodel interface. The "official" version is in the directory 
[controllers/torqueBalancing](/controllers/torqueBalancing), but also other balancing controllers are available. Matlab tools for visualization, inverse 
kinematics, etc. are available in [controllers/tools](/controllers/tools) folder. For more informations about the balancing controller check the relative [README](/controllers/torqueBalancing).

### Simulations
Please note that the simulations of the inverse dynamics results are performed by using the [iDyntree](https://github.com/robotology/iDyntree) visualizer. In order to be able to use it, enable the 
options `IDYNTREE_USES_MATLAB` and `IDYNTREE_USES_IRRLICHT` using `ccmake ./` in iDyntree build directory. If the _irrlicht library_ is not installed on your computer, install it by using 
`sudo apt install libirrlicht-dev`.

## Wbm-toolbox
This folder contains all wrapper and utility functions for Matlab and also the C++ library for the mex-subroutine.

## Tests
To verify regression in the code, when the user modifies it, some regression tests have been implemented in this repository.

### Run tests
Tests use `ctest` infrastructure.
To run the tests on this repository, after compiling the project, just follows these steps:

- Move (`cd`) to the build directory
- Run the `ctest` command. If the project supports multiples configurations (e.g. Xcode), the user also has to specify the configuration with `-C`, thus for example `ctest -C Debug`.

To print out the verbose output of the tests, launch the `ctest` command by adding the `-VV` option.

### Add new tests
To add new tests, just write a script that runs the test and raise and exception
if the test fails (for example using the `assert` command or the [tests/WBAssertEqual.m](tests/WBMAssertEqual.m) function).
Then modify the [tests/WBMTests.m](tests/WBMTests.m) file to call the testing script.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project does not adheres to [Semantic Versioning](http://semver.org/).

## New release: 
### [0.2](https://github.com/robotology/mex-wholebodymodel/compare/v0.2...HEAD) - 2017-01-25

#### Added
- New version of the mexWholeBodyModel (mex-WBM) subroutine for Matlab. Thanks to [@Ganimed](https://github.com/Ganimed), a 
  complete refactory of the toolbox has been done, to speedup and stabilize the call of the kinematic/dynamic functions.

#### New Utility-Functions:
- A lot of utility-functions were added and moved to the utilities-directory in order to be independent from other
  third party libriaries, i.e. transformation functions. For further details have a look into the [+utilities](/mex-wholebodymodel/matlab/utilities/+WBM/+utilities/) directory.

#### Renamed Functions
- **_Wrapper-Function Names:_**
  - wbm_djdq --> wbm_dJdq
  - wbm_getState:
    - Changed order of the output arguments from [qj, xTb, qjDot, vb] to [vqT_b, q_j, v_b, dq_j], where
      qj = q_j, xTb = vqT_b, qjDot = dq_j and vb = v_b.
  - wbm_getWorldFrameFromFixedLink -->  wbm_getWorldFrameFromFixLnk
    - Changed output order:  [world_R_base, world_p_base] --> [wf_p_b, wf_R_b]
- **_Utility-Function Names:_**
  - quaternionDerivative --> WBM.utilities.dquat
    - Changed order of the input argument from (omega,q) to (quat,omega).
    - New optional input arguments: gain (optional), chktol (optional).
  - frame2posrot --> WBM.utilities.frame2posRotm
  - fromSkew2Vector --> WBM.utilities.skewm2vec
  - parametrization --> WBM.utilities.rotm2eulAngVelTF
    - New input arguments: rotm, sequence ('ZYX' or 'ZYZ', default 'ZYX'). 
  - quaternion2dcm --> WBM.utilities.quat2rotm
  - skew --> WBM.utilities.skewm
  - whatname --> WBM.utilities.getJointAnnotationICub
  - wbm_generalisedBiasForces --> wbm_generalizedBiasForces
  - wbm_modelInitialise --> wbm_modelInitialize
  
#### Renamed Key-Names:
- generalised-forces --> generalized-forces
- model-initialise --> model-initialize
- model-initialise-urdf --> model-initialize-urdf
  
#### New Key-Names:
- coriolis-forces
- get-base-state
- gravity-forces
- inverse-dynamics
- transformation-matrix

#### Deleted Functions (outdated):
- fromMesh2sortedVector
- resizeData
- sortedVector2mesh
- visualizeForwardDynamics

#### Future planned renamings:
- Change the variable _q_j_ to _s_j_, where _s_ denotes the _"shape"_ of the robot based on its current
  joint positions _j_.
- The same will be done for the variables _dq_j_ and _ddq_j_, such that _dq_j_ --> _ds_j_ and _ddq_j_ --> _dds_j_.

#### Controllers and WBM-Class
- The controllers in the [controllers](/controllers) directory are adapted to the new wrapper and utility-functions.
- The first [beta release](/mex-wholebodymodel/matlab/wrappers/+WBM) (at the moment undocumented) of the object-oriented version of the Whole Body Model for YARP-based robots is now available for testing purposes.

#### New visualizer
- New toolbox for visualizing the robot simulation based on the iDyntree library. For further information, see [README](https://github.com/robotology/mex-wholebodymodel/blob/master/README.md).

### Using the new functions:
- To avoid conflicts with other libraries, the new utility-functions and also the Matlab-classes are put into own namespaces.
- For applying these new functions the namespace must be specified at first at top of the Matlab procedure or function with
  `import WBM.utilities.<function_name>`, or `import WBM.utilities.*;` to load **all** utility functions, or with
  a direct call of the function with the full namespace path `WBM.utilities.<function_name>`.

## Previous (alpha) version:
### [0.1.1]
[0.1.1]: https://github.com/robotology/mex-wholebodymodel/compare/v0.1...HEAD
#### Added
- All the functions take in input the complete state, including the base pose with respect to the world, 
  see [Issue 27](https://github.com/robotology/mex-wholebodymodel/issues/27).
- Please note that a bug occured before [Pull Request](https://github.com/robotology/mex-wholebodymodel/pull/57) that inverted
  the rotation passed (when passing the full state) to the `wbm_forwardKinematics`,`wbm_massMatrix`, `wbm_jacobian`, `wbm_centrodialMomentum`, `wbm_djdq`. 

#### Removed
- Removed `wbm_setWorldLink`, its behavior can be emulated with the `wbm_getWorldFrameFromFixedLink` and the new interfaces that take the complete
  state as input. 
  See [Issue 43](https://github.com/robotology/mex-wholebodymodel/issues/43) and [Pull Request 52](https://github.com/robotology/mex-wholebodymodel/pull/52).

### [0.1] - 2014-11-15
[0.1]: https://github.com/robotology/mex-wholebodymodel/compare/9fe87c...v0.1
#### Added
- First version of mex-wholebodymodel, compatible with wholebodyinterface v0.1. 


src/timeTest.m
%addpath('./../build');
%Running a 10,000 run trial
totTime = 0;numRuns = 10000;

refLink1 = 'r_hip_1';
refLink2 = 'l_hip_1';

% number of joint for iCubGenova03
nj       = 25;   
% number of joint for iCubHeidelberg01
% nj       = 15;   


fprintf('Starting normal mode trial \n-------------------------- \n');

tic;
wholeBodyModel('model-initialise');
initTime = toc();

fprintf('Initialisation time : %e secs\n',initTime);
fprintf('Num Trials : %d \n Starting Trial...\n',numRuns);
tic;

qj = rand(nj,1);
xTb = wholeBodyModel('forward-kinematics',qj,refLink1);

for i = 1:numRuns
    
    %Setting State to random values
    qj = rand(nj,1);dqj = rand(nj,1);dxb = rand(6,1);
    
    %Mex-WholeBodyModel Components
    M = wholeBodyModel('mass-matrix',qj);    
    H = wholeBodyModel('generalised-forces',qj,dqj,dxb);    
    DjDq1 = wholeBodyModel('djdq',qj,dqj,dxb,refLink1);
    DjDq2 = wholeBodyModel('djdq',qj,dqj,dxb,refLink2);    
    J1 = wholeBodyModel('jacobian',qj,refLink1);J2 = wholeBodyModel('jacobian',qj,refLink2);
    J3 = wholeBodyModel('jacobian',qj,refLink1);J4 = wholeBodyModel('jacobian',qj,refLink2);
end
totTime = toc();
%Benchmarks
fprintf('Normal-Mode Trial Total Time : %f secs \n',totTime);
fprintf('Normal-Mode Trial Average Time : %e secs\n',totTime/numRuns);

fprintf('\n\nStarting optimised mode trial \n-------------------------- \n');

totTime = 0;numRuns = 10000;
tic;
wholeBodyModel('model-initialise');
initTime = toc();
fprintf('Initialisation time : %e secs \n Starting Trial...\n ',initTime);

refLink1 = 'r_hip_1';
refLink2 = 'l_hip_1';

tic;
for i = 1:numRuns
    
    %Setting State to random values
    qj = rand(nj,1);dqj = rand(nj,1);dxb = rand(6,1);
    
    wholeBodyModel('update-state',qj,dqj,dxb);
    
    %Mex-WholeBodyModel Components
    M = wholeBodyModel('mass-matrix');    
    H = wholeBodyModel('generalised-forces');    
    DjDq1 = wholeBodyModel('djdq',refLink1);
    DjDq2 = wholeBodyModel('djdq',refLink2);    
    J1 = wholeBodyModel('jacobian',refLink1);J2 = wholeBodyModel('jacobian',refLink2);
    J3 = wholeBodyModel('jacobian',refLink1);J4 = wholeBodyModel('jacobian',refLink2);
end
totTime = toc();
%Benchmarks
fprintf('Optimised-Mode Trial Total Time : %f secs\n',totTime);
fprintf('Optimised-Mode Trial Average Time : %e secs \n',totTime/numRuns);
